### PR TITLE
Fix #178: Make MNT{4,6}-753, cp6_782 tests run conditionally

### LINKED ISSUE #178: Make CI conditional on which curves are changed
Right now MNT{4,6}-753 take a long time to complete execution, slowing down CI overall by a large amount. We should make a more intelligent CI that only runs tests for curves if either
(a) the dependencies have changed, or
(b) the curve code has changed.

### LINKED ISSUE #180: Run cp6_782 conditionally like MNT{4,6}-753

### PR DESCRIPTION
**Pull request by z-tech**
## Make MNT{4,6}-753, cp6_782 tests run conditionally

Fixes #178
Fixes #180 

What does this PR do?

- updates ci.yml so that tests for MNT{4,6}-753 and cp6_782 are run only when either:
  - a) relevant source code changes
  - b) dependencies are updated
  
- I sanity checked the following conditions:
  - code was updated
  - code was not updated
  - deps were updated
  - deps were not updated

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests n/a
- [ ] Updated relevant documentation in the code n/a
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md` n/a
- [ ] Re-reviewed `Files changed` in the Github PR explorer ?

### CODE REVIEW

**Comment by Pratyush** [comment]
This is a good start! Nice find on the action. However as things stand this is a bit too aggressive, because it doesn't handle when the dependencies change. I think the way to handle that would be to check the current `Cargo.lock` against the previous `Cargo.lock`. To perform the comparison we would need to cache the old `Cargo.lock`.

**Review by Pratyush — APPROVED** [review]

**Comment by Pratyush** [comment]
Hm the CI is failing with the following error:

```
[Invalid workflow file: .github/workflows/ci.yml#L124](https://github.com/arkworks-rs/curves/actions/runs/6275003357/workflow)
The workflow is not valid. .github/workflows/ci.yml (Line: 124, Col: 9): Unrecognized named-value: 'steps'. Located at position 1 within expression: steps.filter.outputs.mnt4_753_is_updated == 'true' .github/workflows/ci.yml (Line: 137, Col: 9): Unrecognized named-value: 'steps'. Located at position 1 within expression: steps.filter.outputs.mnt6_753_is_updated == 'true'
```

**Comment by z-tech** [comment]
> This is a good start! Nice find on the action. However as things stand this is a bit too aggressive, because it doesn't handle when the dependencies change. I think the way to handle that would be to check the current `Cargo.lock` against the previous `Cargo.lock`. To perform the comparison we would need to cache the old `Cargo.lock`.

Okay cool, I agree. I'll look into caching Cargo.lock and debug that other issue.

Did we want to do conditional for _all_ test suites? (not just mnt4_753, mnt6_753)

**Comment by z-tech** [comment]
Okay sweet I think this does what we want LMK

**Comment by Pratyush** [comment]
Looks great! Maybe the only diff would be to update the other `cache` action invocation (line 52) to v3 also.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -115,27 +116,109 @@ jobs:
...
+
+      - name: Check if deps updated
+        id: deps-updated
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+            target
+          key: mnt4_753-deps-updated-${{ runner.os }}-${{ hashFiles('**/Cargo.lock') }}
```
Comment (Pratyush) [inline .github/workflows/ci.yml:142]: I think we can avoid the `mnt4_753-deps-updated-` prefix, as the `Cargo.lock` file is shared across the entire workspace. So we'd be unnecessarily creating an additional cache entry and possibly causing false cache misses if we use this key. Ditto for the other curves.

Comment (z-tech) [reply .github/workflows/ci.yml]: That last change didn't result in "cache hits" like I expected...

Does `cargo generate-lockfile` need to be added before the cache invocation at line 53?

**Comment by Pratyush** [comment]
Mostly looks good to me! A minor optimization that we could do would be to simply check to see if the cache exists, instead of downloading it also. See https://github.com/actions/cache#inputs for how to do this. Thanks for the great work!

**Comment by z-tech** [comment]
> Mostly looks good to me! A minor optimization that we could do would be to simply check to see if the cache exists, instead of downloading it also. See https://github.com/actions/cache#inputs for how to do this. Thanks for the great work!

Sweet! Thanks for the help! I made a little note: https://github.com/arkworks-rs/curves/issues/181

**Pull request merged**